File: packages/health_connector_hc_android/android/src/test/kotlin/com/phamtunglam/health_connector_hc_android/unit_tests/mappers/health_record_mappers/ExerciseSessionEventMapperTest.kt
package com.phamtunglam.health_connector_hc_android.unit_tests.mappers.health_record_mappers

import androidx.health.connect.client.records.ExerciseLap
import androidx.health.connect.client.records.ExerciseSegment
import androidx.health.connect.client.units.Length
import com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.ExerciseSegmentTypeDto
import com.phamtunglam.health_connector_hc_android.pigeon.ExerciseSessionLapEventDto
import com.phamtunglam.health_connector_hc_android.pigeon.ExerciseSessionSegmentEventDto
import io.kotest.matchers.shouldBe
import java.time.Instant
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test

@DisplayName("ExerciseSessionEventMapper")
class ExerciseSessionEventMapperTest {

    private companion object {
        const val TEST_START_TIME = 1609459200000L // 2021-01-01 00:00:00 UTC
        const val TEST_END_TIME = 1609459260000L // 2021-01-01 00:01:00 UTC
        const val TEST_REPETITIONS = 10
        const val TEST_DISTANCE_METERS = 100.0
    }

    @Test
    @DisplayName("GIVEN ExerciseSegment → WHEN toDto → THEN converts correctly")
    fun whenExerciseSegmentToDto_thenConvertsCorrectly() {
        // Given
        val segment = ExerciseSegment(
            startTime = Instant.ofEpochMilli(TEST_START_TIME),
            endTime = Instant.ofEpochMilli(TEST_END_TIME),
            segmentType = ExerciseSegment.EXERCISE_SEGMENT_TYPE_RUNNING,
            repetitions = TEST_REPETITIONS,
        )

        // When
        val dto = segment.toDto()

        // Then
        dto.startTime shouldBe TEST_START_TIME
        dto.endTime shouldBe TEST_END_TIME
        dto.segmentType shouldBe ExerciseSegmentTypeDto.RUNNING
        dto.repetitions shouldBe TEST_REPETITIONS.toLong()
    }

    @Test
    @DisplayName("GIVEN ExerciseSegment with 0 repetitions → WHEN toDto → THEN repetitions is null")
    fun whenExerciseSegmentWithZeroRepetitionsToDto_thenRepetitionsIsNull() {
        // Given
        val segment = ExerciseSegment(
            startTime = Instant.ofEpochMilli(TEST_START_TIME),
            endTime = Instant.ofEpochMilli(TEST_END_TIME),
            segmentType = ExerciseSegment.EXERCISE_SEGMENT_TYPE_WALKING,
            repetitions = 0,
        )

        // When
        val dto = segment.toDto()

        // Then
        dto.repetitions shouldBe null
    }

    @Test
    @DisplayName("GIVEN ExerciseLap → WHEN toDto → THEN converts correctly")
    fun whenExerciseLapToDto_thenConvertsCorrectly() {
        // Given
        val lap = ExerciseLap(
            startTime = Instant.ofEpochMilli(TEST_START_TIME),
            endTime = Instant.ofEpochMilli(TEST_END_TIME),
            length = Length.meters(TEST_DISTANCE_METERS),
        )

        // When
        val dto = lap.toDto()

        // Then
        dto.startTime shouldBe TEST_START_TIME
        dto.endTime shouldBe TEST_END_TIME
        dto.distanceMeters shouldBe TEST_DISTANCE_METERS
    }

    @Test
    @DisplayName(
        "GIVEN ExerciseSessionSegmentEventDto → WHEN toHealthConnect → THEN converts correctly",
    )
    fun whenExerciseSessionSegmentEventDtoToHealthConnect_thenConvertsCorrectly() {
        // Given
        val dto = ExerciseSessionSegmentEventDto(
            startTime = TEST_START_TIME,
            endTime = TEST_END_TIME,
            segmentType = ExerciseSegmentTypeDto.SQUAT,
            repetitions = TEST_REPETITIONS.toLong(),
        )

        // When
        val segment = dto.toHealthConnect()

        // Then
        segment.startTime shouldBe Instant.ofEpochMilli(TEST_START_TIME)
        segment.endTime shouldBe Instant.ofEpochMilli(TEST_END_TIME)
        segment.segmentType shouldBe ExerciseSegment.EXERCISE_SEGMENT_TYPE_SQUAT
        segment.repetitions shouldBe TEST_REPETITIONS
    }

    @Test
    @DisplayName(
        "GIVEN ExerciseSessionSegmentEventDto with null repetitions → " +
            "WHEN toHealthConnect → THEN repetitions is 0",
    )
    fun whenExerciseSessionSegmentEventWithNullRepetitionsToHealthConnect_thenRepetitionsIsZero() {
        // Given
        val dto = ExerciseSessionSegmentEventDto(
            startTime = TEST_START_TIME,
            endTime = TEST_END_TIME,
            segmentType = ExerciseSegmentTypeDto.BIKING,
            repetitions = null,
        )

        // When
        val segment = dto.toHealthConnect()

        // Then
        segment.repetitions shouldBe 0
    }

    @Test
    @DisplayName(
        "GIVEN ExerciseSessionLapEventDto → WHEN toHealthConnect → THEN converts correctly",
    )
    fun whenExerciseSessionLapEventDtoToHealthConnect_thenConvertsCorrectly() {
        // Given
        val dto = ExerciseSessionLapEventDto(
            startTime = TEST_START_TIME,
            endTime = TEST_END_TIME,
            distanceMeters = TEST_DISTANCE_METERS,
        )

        // When
        val lap = dto.toHealthConnect()

        // Then
        lap.startTime shouldBe Instant.ofEpochMilli(TEST_START_TIME)
        lap.endTime shouldBe Instant.ofEpochMilli(TEST_END_TIME)
        lap.length shouldBe Length.meters(TEST_DISTANCE_METERS)
    }
}

File: packages/health_connector_hc_android/test/unit_tests/src/mappers/health_record_mappers/exercise/exercise_session_event_mapper_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:health_connector_core/health_connector_core_internal.dart';
import 'package:health_connector_hc_android/src/mappers/health_record_mappers/exercise/exercise_session_event_mapper.dart';
import 'package:health_connector_hc_android/src/pigeon/health_connector_hc_android_api.g.dart';

void main() {
  group('ExerciseSessionEventMapper', () {
    test('throws UnsupportedError for ExerciseSessionStateTransitionEvent', () {
      final event = ExerciseSessionStateTransitionEvent(
        time: DateTime.now(),
        type: ExerciseSessionStateTransitionType.pause,
      );

      expect(event.toDto, throwsUnsupportedError);
    });

    test('throws UnsupportedError for ExerciseSessionMarkerEvent', () {
      final event = ExerciseSessionMarkerEvent(time: DateTime.now());

      expect(event.toDto, throwsUnsupportedError);
    });

    test('converts ExerciseSessionLapEvent to/from DTO', () {
      final startTime = DateTime(2023, 1, 1, 10).toUtc();
      final endTime = DateTime(2023, 1, 1, 10, 30).toUtc();
      final event = ExerciseSessionLapEvent(
        startTime: startTime,
        endTime: endTime,
        distance: const Length.meters(100),
      );

      final dto = event.toDto() as ExerciseSessionLapEventDto;
      expect(dto.startTime, startTime.millisecondsSinceEpoch);
      expect(dto.endTime, endTime.millisecondsSinceEpoch);
      expect(dto.distanceMeters, 100.0);

      final domain = dto.toDomain() as ExerciseSessionLapEvent;
      expect(domain.startTime, startTime);
      expect(domain.endTime, endTime);
      expect(domain.distance, const Length.meters(100));
    });

    test('converts ExerciseSessionSegmentEvent to/from DTO', () {
      final startTime = DateTime(2023, 1, 1, 10).toUtc();
      final endTime = DateTime(2023, 1, 1, 10, 30).toUtc();
      final event = ExerciseSessionSegmentEvent(
        startTime: startTime,
        endTime: endTime,
        segmentType: ExerciseSegmentType.running,
        repetitions: 10,
      );

      final dto = event.toDto() as ExerciseSessionSegmentEventDto;
      expect(dto.startTime, startTime.millisecondsSinceEpoch);
      expect(dto.endTime, endTime.millisecondsSinceEpoch);
      expect(dto.segmentType, ExerciseSegmentTypeDto.running);
      expect(dto.repetitions, 10);

      final domain = dto.toDomain() as ExerciseSessionSegmentEvent;
      expect(domain.startTime, startTime);
      expect(domain.endTime, endTime);
      expect(domain.segmentType, ExerciseSegmentType.running);
      expect(domain.repetitions, 10);
    });
  });
}
